internal/services/v1/relationships_test.go

```diff
@@ -331,6 +336,56 @@ func TestWriteRelationships(t *testing.T) {
 	require.ErrorIs(err, io.EOF)
 }
 
+func TestWriteCaveatedRelationships(t *testing.T) {
+	req := require.New(t)
+
+	conn, cleanup, ds, _ := testserver.NewTestServer(req, 0, memdb.DisableGC, true, tf.StandardDatastoreWithData)
+	client := v1.NewPermissionsServiceClient(conn)
+	t.Cleanup(cleanup)
+
+	toWrite := tuple.MustParse("document:totallynew#parent@folder:plans")
+	caveatCtx, err := structpb.NewStruct(map[string]any{"a": 1, "b": 2})
+	req.NoError(err)
+	toWrite.Caveat = &core.ContextualizedCaveat{
+		CaveatName: "test",
+		Context:    caveatCtx,
+	}
+
+	toWrite.Caveat.Context = caveatCtx
+	relWritten := tuple.MustToRelationship(toWrite)
+	writeReq := &v1.WriteRelationshipsRequest{
+		Updates: []*v1.RelationshipUpdate{{
+			Operation:    v1.RelationshipUpdate_OPERATION_CREATE,
+			Relationship: relWritten,
+		}},
+	}
+
+	// Should fail due to non-existing caveat
+	ctx := context.Background()
+	_, err = client.WriteRelationships(ctx, writeReq)
+	grpcutil.RequireStatus(t, codes.FailedPrecondition, err)
+
+	// TODO(vroldanbet) temporarily use Datastore to write caveat until we expose it via Schema API
+	_, err = ds.ReadWriteTx(ctx, func(ctx context.Context, transaction datastore.ReadWriteTransaction) error {
+		caveatStorer, ok := transaction.(datastore.CaveatStorer)
+		req.True(ok, "expected datastore to implement datastore.CaveatStorer")
+		return caveatStorer.WriteCaveats([]*core.Caveat{{
+			Name:       "test",
+			Expression: []byte{},
+		}})
+	})
+	req.NoError(err)
+
+	// should succeed
+	resp, err := client.WriteRelationships(context.Background(), writeReq)
+	req.NoError(err)
+
+	// read relationship back
+	relRead := readFirst(req, client, resp.WrittenAt, relWritten)
+	foundDiff := cmp.Diff(relWritten, relRead, protocmp.Transform())
```

Comment:
Could also use `proto.Equal`

Reply:
That's not the convention used in the codebase. It uses `indexID` because it's the "ID" column for `hashicorp/go-memdb`, which is required. You can think of it as the "PK" constraint